File: No_802/Find Eventual Safe States.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        vector<int> color(graph.size(), 0); // 0: not visited, 1: safe, 2: unsafe
        vector<int> ret;
        for(int i=0; i<graph.size(); i++) {
            if(DFS(graph, i, color)) {
                ret.push_back(i);
            }
        }
        
        return ret;
    }
    
    bool DFS(vector<vector<int>>& graph, int n, vector<int>& color) {
        if(color[n] == 1) return true;
        if(color[n] == 2) return false;
        
        color[n] = 2; // default unsafe
        for(int e : graph[n]) { // check if every path starting from n leads to a safe node
            if(!DFS(graph, e, color)) return false;
        }
        
        color[n] = 1;
        return true;
    }
};

File: No_802/Find Eventual Safe States.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        def DFS(graph, color, n) -> bool:
            if color[n] == 1:
                return True
            if color[n] == 2:
                return False
            
            color[n] = 2
            for i in graph[n]:
                if not DFS(graph, color, i):
                    return False
            
            color[n] = 1
            return True
            
        n = len(graph)
        ret = []
        color = [0] * n #  0: not visited, 1: safe, 2: unsafe
        for i in range(n):
            if DFS(graph, color, i):
                ret.append(i)
            
        return ret